connect-inject/health_check_handler.go

```diff
@@ -0,0 +1,317 @@
+type HealthCheckHandler struct {
+	Log                log.Logger
+	AclConfig          api.NamespaceACLConfig
+	Client             *api.Client
+	HFlags             *flags.HTTPFlags
+	Clientset          kubernetes.Interface
+	ConsulClientScheme string
+	ConsulPort         string
+}
+
+// getConsulHealthCheckID deterministically generates a health check ID that will be unique to the Agent
+// where the health check is registered and deregistered.
+func (t *HealthCheckHandler) getConsulHealthCheckID(pod *corev1.Pod) string {
+	return fmt.Sprintf("%s_%s_kubernetes-health-check-ttl", pod.Namespace, t.getConsulServiceID(pod))
+}
+
+// return the serviceID of the connect service
+func (t *HealthCheckHandler) getConsulServiceID(pod *corev1.Pod) string {
+	return pod.Name + "-" + pod.Annotations[annotationService]
+	// TODO: I'd like to have this put into an annotation by this point, unfortunately it cannot
+	// be injected by the webhook because pod.Name isnt determined by then
+	//return pod.Annotations[annotationServiceID]
+}
+
+// updateConsulClient updates the Consul Client metadata to point to a new hostIP:port
+// which is the IP of the host that the Pod runs on, in order to make Agent calls locally
+// for health check registration/deregistration.
+func (t *HealthCheckHandler) updateConsulClient(pod *corev1.Pod) error {
+	var err error
+	// We use the whole schema including http/https here because that is normally set in the
+	// CONSUL_HTTP_ADDR environment variable, without https the client will default to http even if the port
+	// is 8501
+	httpFmt := "http"
+	if t.ConsulPort == "8501" {
+		httpFmt = "https"
+	}
+	newAddr := fmt.Sprintf("%s://%s:%s", httpFmt, pod.Status.HostIP, t.ConsulPort)
+	t.HFlags.SetAddress(newAddr)
+
+	// Set client api to point to the new host IP
+	// TODO: APIClient() reads the ENV, see if we can avoid this for perf
+	t.Client, err = t.HFlags.APIClient()
+	if err != nil {
+		t.Log.Error("unable to get Consul API Client for address %s: %s", newAddr, err)
+		t.Client = nil
+	}
+	t.Log.Info("setting consul client to the following agent: %v", newAddr)
+	return err
+}
+
+// registerConsulHealthCheck registers a Failing TTL health check for the service on this Agent.
+// The Agent is local to the Pod which has failed a health check.
+// This has the effect of marking the endpoint unhealthy for Consul service mesh traffic.
+func (t *HealthCheckHandler) registerConsulHealthCheck(pod *corev1.Pod, consulHealthCheckID, serviceID, initialStatus, reason string) error {
+	t.Log.Debug("registerConsulHealthCheck, %v %v", consulHealthCheckID, serviceID)
+	// There is a chance of a race between when the Pod is transitioned to healthy by k8s and when we've initially
+	// completed the registration of the service with the Consul Agent on this node. Retry a few times to be sure
+	// that the service does in fact exist, otherwise it will return 500 from Consul API.
+	retries := 0
+	var err error
+	err = backoff.Retry(func() error {
```

Comment:
I'm a little worried about this retry loop. Right now everything is happening synchronously. If registration takes a long time or network traffic is slow, then this could block for 10s during which time no other updates will be made to any other pod statuses.

We could have this run in a separate goroutine or have a separate queue per consul node for api calls to that agent?

Reply:
I believe this may no longer be needed actually, I did a little thought experiment today on this and since we are no longer triggering on the PodCreate but rather the PodRunning prior to reaching this handle it should be almost a certainty that the init_container has already completed running which means the service register has completed.
I still need to test this theory though, but if it holds I think we can remove it altogether and utilize the re-queue retry logic instead.